# WEB SCRAPING

First we need to import all the necessary packages.

In [4]:
pip install geopy

     |████████████████████████████████| 112kB 5.1MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [6]:
pip install lxml

     |████████████████████████████████| 5.5MB 8.1MB/s eta 0:00:01��██▍  | 5.1MB 8.1MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [7]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print("Libraries imported.")
import requests
import lxml.html as lh
import pandas as pd
import numpy as np

Libraries imported.


Then go to the url and right click on the element to scrape.Click on inspect and find out the tag containing the dataset. 

In [8]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
#Create a handle, page, to handle the contents of the website
page = requests.get(url)
#Store the contents of the website under doc
doc = lh.fromstring(page.content)
#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')

In [9]:
#Check the length of the first 12 rows
[len(T) for T in tr_elements[:12]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

Now we need to get the first row of the table that is the header row and print it.



In [10]:
tr_elements = doc.xpath('//tr')
#Create empty list
col=[]
i=0
#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print ('%d:"%s"'%(i,name))
    col.append((name,[]))

1:"Postal Code
"
2:"Borough
"
3:"Neighborhood
"


Append the data from the columns row after row.

In [11]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    T=tr_elements[j]
    
    #If row is not of size 3, the //tr data is not from our table 
    if len(T)!=3:
        break
    
    #i is the index of our column
    i=0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        data=t.text_content() 
        #Check if row is empty
        if i>0:
        #Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1

We can check if the data from all the columns have been successfully obtained by checking their length.

In [12]:
[len(C) for (title,C) in col]

[181, 181, 181]

Now we are sure of the length of the column.So we can convert the data into a dataframe using Pandas.

In [13]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)
df.head()

Postal Code\n           Borough\n               Neighborhood\n
0         M1A\n      Not assigned\n               Not assigned\n
1         M2A\n      Not assigned\n               Not assigned\n
2         M3A\n        North York\n                  Parkwoods\n
3         M4A\n        North York\n           Victoria Village\n
4         M5A\n  Downtown Toronto\n  Regent Park, Harbourfront\n

In [14]:
print("The number of rows in the dataframe are {}".format(df.shape[0]))

The number of rows in the dataframe are 181


Replace Not assigned in the borough column to NaN so that they can be easily removed.

In [15]:
df.replace('Not assigned\n',np.nan,inplace=True)
df.dropna(subset=['Borough\n'],axis=0,inplace=True)
df.reset_index(drop=True,inplace=True)

In [16]:
df.head()

Postal Code\n           Borough\n  \
0         M3A\n        North York\n   
1         M4A\n        North York\n   
2         M5A\n  Downtown Toronto\n   
3         M6A\n        North York\n   
4         M7A\n  Downtown Toronto\n   

                                  Neighborhood\n  
0                                    Parkwoods\n  
1                             Victoria Village\n  
2                    Regent Park, Harbourfront\n  
3             Lawrence Manor, Lawrence Heights\n  
4  Queen's Park, Ontario Provincial Government\n

In [17]:
print("The number of rows in the dataframe are {}".format(df.shape[0]))

The number of rows in the dataframe are 104


After removing the rows where Borough is unavailable ,there are now 104 rows instead of 181.

But my data needs to be cleaned up a bit removing all whitespace characters.

In [18]:
df.replace(r'\s', '', regex = True, inplace = True)

In [19]:
df.head()

Postal Code\n        Borough\n                           Neighborhood\n
0           M3A        NorthYork                                Parkwoods
1           M4A        NorthYork                          VictoriaVillage
2           M5A  DowntownToronto                  RegentPark,Harbourfront
3           M6A        NorthYork            LawrenceManor,LawrenceHeights
4           M7A  DowntownToronto  Queen'sPark,OntarioProvincialGovernment

Cleaning the row headers:

In [20]:
df.columns=['Postal Code','Borough','Neighborhood']
df.head()

Postal Code          Borough                             Neighborhood
0         M3A        NorthYork                                Parkwoods
1         M4A        NorthYork                          VictoriaVillage
2         M5A  DowntownToronto                  RegentPark,Harbourfront
3         M6A        NorthYork            LawrenceManor,LawrenceHeights
4         M7A  DowntownToronto  Queen'sPark,OntarioProvincialGovernment

Group the neighborhoods in the same borough.

In [21]:
# group neighborhoods in the same borough
df_grouped = df.groupby(["Postal Code", "Borough"], as_index=False).agg(lambda x: ", ".join(x))
df_grouped.head()

Postal Code              Borough                       Neighborhood
0              Canadianpostalcodes                                   
1         M1B          Scarborough                      Malvern,Rouge
2         M1C          Scarborough  RougeHill,PortUnion,HighlandCreek
3         M1E          Scarborough     Guildwood,Morningside,WestHill
4         M1G          Scarborough                             Woburn

In [20]:
df_grouped=df_grouped.iloc[1:,:]
df_grouped.reset_index(drop=True,inplace=True)

In [22]:
df_grouped.head()

Postal Code              Borough                       Neighborhood
0              Canadianpostalcodes                                   
1         M1B          Scarborough                      Malvern,Rouge
2         M1C          Scarborough  RougeHill,PortUnion,HighlandCreek
3         M1E          Scarborough     Guildwood,Morningside,WestHill
4         M1G          Scarborough                             Woburn

In [22]:
# for Neighborhood=NaN, make the value the same as Borough
for index, row in df_grouped.iterrows():
    if row["Neighborhood"] == np.nan:
        row["Neighborhood"] = row["Borough"]
        
df_grouped.head()

Postal Code      Borough                       Neighborhood
0         M1B  Scarborough                      Malvern,Rouge
1         M1C  Scarborough  RougeHill,PortUnion,HighlandCreek
2         M1E  Scarborough     Guildwood,Morningside,WestHill
3         M1G  Scarborough                             Woburn
4         M1H  Scarborough                          Cedarbrae

In [23]:
print("The number of rows in the dataframe are {}".format(df.shape[0]))

The number of rows in the dataframe are 104


# Converting the postal codes to latitudes and longitudes

In [24]:
csv_path="http://cocl.us/Geospatial_data"
df_coord=pd.read_csv(csv_path)
df_coord.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Now we need to merge the two dataframes making sure that the Postal Codes match.

In [25]:
# merge two table on the column "PostalCode"
df = df_grouped.merge(df_coord, on="Postal Code", how="left")
df.head()

Postal Code              Borough                       Neighborhood  \
0              Canadianpostalcodes                                      
1         M1B          Scarborough                      Malvern,Rouge   
2         M1C          Scarborough  RougeHill,PortUnion,HighlandCreek   
3         M1E          Scarborough     Guildwood,Morningside,WestHill   
4         M1G          Scarborough                             Woburn   

    Latitude  Longitude  
0        NaN        NaN  
1  43.806686 -79.194353  
2  43.784535 -79.160497  
3  43.763573 -79.188711  
4  43.770992 -79.216917

### Use geopy library to get the latitude and longitude values of Toronto City.

In [41]:
address = 'Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


### Create a map of Toronto with neighborhoods superimposed on top

In [42]:
# create map of Toronto using latitude and longitude values
map_tor = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'],df['Longitude'],df['Borough'],df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_tor)  
    
map_tor

In [43]:
# filter borough names that contain the word Toronto
borough_names = list(df.Borough.unique())
borough_toronto = []
for x in borough_names:
    if "toronto" in x.lower():
        borough_toronto.append(x)        
borough_toronto

['EastToronto', 'CentralToronto', 'DowntownToronto', 'WestToronto']

In [44]:
df = df[df['Borough'].isin(borough_toronto)].reset_index(drop=True)
df.head()

Postal Code         Borough                Neighborhood   Latitude  \
0         M4E     EastToronto                  TheBeaches  43.676357   
1         M4K     EastToronto   TheDanforthWest,Riverdale  43.679557   
2         M4L     EastToronto  IndiaBazaar,TheBeachesWest  43.668999   
3         M4M     EastToronto              StudioDistrict  43.659526   
4         M4N  CentralToronto                LawrencePark  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

### Use the Foursquare API to explore the neighborhoods

In [45]:
# define Foursquare Credentials and Version
CLIENT_ID = '3GSS143JROHTSV3BTRXDZW1F4LOZ00AIP3UKWZFVEYNZM2JE' # your Foursquare ID
CLIENT_SECRET = 'TRTTM4GAEJ2JIKYXN5EVKX1UPOZR0FDYPEIUR4UXCIRIZQIV' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 3GSS143JROHTSV3BTRXDZW1F4LOZ00AIP3UKWZFVEYNZM2JE
CLIENT_SECRET:TRTTM4GAEJ2JIKYXN5EVKX1UPOZR0FDYPEIUR4UXCIRIZQIV


To get the top 100 venues within 500 metres

In [46]:
radius = 500
LIMIT = 100
venues = []
for lat, long, post, borough, neighborhood in zip(df['Latitude'],df['Longitude'],df['Postal Code'],df['Borough'],df['Neighborhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    results = requests.get(url).json()["response"]['groups'][0]['items']
    for venue in results:
        venues.append((
            post, 
            borough,
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [47]:
# convert the venues list into a new DataFrame
df_of_venues = pd.DataFrame(venues)
df_of_venues.head()

0            1                          2          3          4  \
0  M4E  EastToronto                 TheBeaches  43.676357 -79.293031   
1  M4E  EastToronto                 TheBeaches  43.676357 -79.293031   
2  M4E  EastToronto                 TheBeaches  43.676357 -79.293031   
3  M4E  EastToronto                 TheBeaches  43.676357 -79.293031   
4  M4K  EastToronto  TheDanforthWest,Riverdale  43.679557 -79.352188   

                                    5          6          7                  8  
0                   Glen Manor Ravine  43.676821 -79.293942              Trail  
1  The Big Carrot Natural Food Market  43.678879 -79.297734  Health Food Store  
2                 Grover Pub and Grub  43.679181 -79.297215                Pub  
3                       Upper Beaches  43.680563 -79.292869       Neighborhood  
4                       MenEssentials  43.677820 -79.351265     Cosmetics Shop

In [48]:
df_of_venues.columns = ['Postal Code', 'Borough', 'Neighborhood', 'Borough Latitude', 'Borough Longitude', 'Venue Name', 'Venue Latitude', 'Venue Longitude', 'Venue Category']

In [49]:
df_of_venues.head()

Postal Code      Borough               Neighborhood  Borough Latitude  \
0         M4E  EastToronto                 TheBeaches         43.676357   
1         M4E  EastToronto                 TheBeaches         43.676357   
2         M4E  EastToronto                 TheBeaches         43.676357   
3         M4E  EastToronto                 TheBeaches         43.676357   
4         M4K  EastToronto  TheDanforthWest,Riverdale         43.679557   

   Borough Longitude                          Venue Name  Venue Latitude  \
0         -79.293031                   Glen Manor Ravine       43.676821   
1         -79.293031  The Big Carrot Natural Food Market       43.678879   
2         -79.293031                 Grover Pub and Grub       43.679181   
3         -79.293031                       Upper Beaches       43.680563   
4         -79.352188                       MenEssentials       43.677820   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

In [50]:
df_of_venues.shape

(1625, 9)

In [51]:
toronto_grouped = df_of_venues.groupby(["Postal Code", "Borough", "Neighborhood"]).mean().reset_index()
toronto_grouped

Postal Code          Borough  \
0          M4E      EastToronto   
1          M4K      EastToronto   
2          M4L      EastToronto   
3          M4M      EastToronto   
4          M4N   CentralToronto   
5          M4P   CentralToronto   
6          M4R   CentralToronto   
7          M4S   CentralToronto   
8          M4T   CentralToronto   
9          M4V   CentralToronto   
10         M4W  DowntownToronto   
11         M4X  DowntownToronto   
12         M4Y  DowntownToronto   
13         M5A  DowntownToronto   
14         M5B  DowntownToronto   
15         M5C  DowntownToronto   
16         M5E  DowntownToronto   
17         M5G  DowntownToronto   
18         M5H  DowntownToronto   
19         M5J  DowntownToronto   
20         M5K  DowntownToronto   
21         M5L  DowntownToronto   
22         M5N   CentralToronto   
23         M5P   CentralToronto   
24         M5R   CentralToronto   
25         M5S  DowntownToronto   
26         M5T  DowntownToronto   
27         M5V  DowntownToronto   
28         M5W  DowntownToronto   
29         M5X  DowntownToronto   
30         M6G  DowntownToronto   
31         M6H      WestToronto   
32         M6J      WestToronto   
33         M6K      WestToronto   
34         M6P      WestToronto   
35         M6R      WestToronto   
36         M6S      WestToronto   
37         M7A  DowntownToronto   
38         M7Y      EastToronto   

                                         Neighborhood  Borough Latitude  \
0                                          TheBeaches         43.676357   
1                           TheDanforthWest,Riverdale         43.679557   
2                          IndiaBazaar,TheBeachesWest         43.668999   
3                                      StudioDistrict         43.659525   
4                                        LawrencePark         43.728020   
5                                     DavisvilleNorth         43.712751   
6                       NorthTorontoWest,LawrencePark         43.715383   
7                                          Davisville         43.704324   
8                            MoorePark,SummerhillEast         43.689574   
9   SummerhillWest,Rathnelly,SouthHill,ForestHillS...         43.686412   
10                                           Rosedale         43.679563   
11                           St.JamesTown,Cabbagetown         43.667967   
12                                 ChurchandWellesley         43.665860   
13                            RegentPark,Harbourfront         43.654260   
14                             GardenDistrict,Ryerson         43.657162   
15                                       St.JamesTown         43.651494   
16                                         BerczyPark         43.644771   
17                                   CentralBayStreet         43.657952   
18                             Richmond,Adelaide,King         43.650571   
19       HarbourfrontEast,UnionStation,TorontoIslands         43.640816   
20               TorontoDominionCentre,DesignExchange         43.647177   
21                        CommerceCourt,VictoriaHotel         43.648199   
22                                           Roselawn         43.711695   
23            ForestHillNorth&West,ForestHillRoadPark         43.696948   
24                    TheAnnex,NorthMidtown,Yorkville         43.672710   
25                        UniversityofToronto,Harbord         43.662696   
26              KensingtonMarket,Chinatown,GrangePark         43.653206   
27  CNTower,KingandSpadina,RailwayLands,Harbourfro...         43.628947   
28                                        StnAPOBoxes         43.646435   
29                 FirstCanadianPlace,Undergroundcity         43.648429   
30                                           Christie         43.669542   
31                         Dufferin,DovercourtVillage         43.669005   
32                             LittlePortugal,Trinity         43.647927   
33           Brockton,ParkdaleVillage,ExhibitionPlac

In [52]:
toronto_onehot = pd.get_dummies(df_of_venues[['Venue Category']], prefix="", prefix_sep="")

# add postal, borough and neighborhood column back to dataframe
toronto_onehot['PostalCode'] = df_of_venues['Postal Code'] 
toronto_onehot['Borough'] = df_of_venues['Borough'] 
toronto_onehot['Neighborhoods'] = df_of_venues['Neighborhood'] 

# move postal, borough and neighborhood column to the first column
fixed_columns = list(toronto_onehot.columns[-3:]) + list(toronto_onehot.columns[:-3])
toronto_onehot = toronto_onehot[fixed_columns]

print(toronto_onehot.shape)
toronto_onehot.head()

(1625, 235)


PostalCode      Borough              Neighborhoods  Airport  \
0        M4E  EastToronto                 TheBeaches        0   
1        M4E  EastToronto                 TheBeaches        0   
2        M4E  EastToronto                 TheBeaches        0   
3        M4E  EastToronto                 TheBeaches        0   
4        M4K  EastToronto  TheDanforthWest,Riverdale        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Auto Workshop  BBQ Joint  \
0                    0                 0              0          0   
1                    0                 0              0          0   
2                    0                 0              0          0   
3                    0                 0              0          0   
4                    0                 0              0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0           0           0       0     0    0                 0   
1           0           0       0     0    0                 0   
2           0           0       0     0    0                 0   
3           0           0       0     0    0                 0   
4           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0                   0       0           0              0          0         0   
1                   0       0           0              0          0         0   
2                   0       0           0              0          0         0   
3                   0       0           0              0          0         0   
4                   0       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Camera Store  Candy Store  Caribbean Restaurant  \
0                          0             0            0     

In [54]:
toronto_grouped = toronto_onehot.groupby(["PostalCode", "Borough", "Neighborhoods"]).mean().reset_index()

print(toronto_grouped.shape)
toronto_grouped

(39, 235)


PostalCode          Borough  \
0         M4E      EastToronto   
1         M4K      EastToronto   
2         M4L      EastToronto   
3         M4M      EastToronto   
4         M4N   CentralToronto   
5         M4P   CentralToronto   
6         M4R   CentralToronto   
7         M4S   CentralToronto   
8         M4T   CentralToronto   
9         M4V   CentralToronto   
10        M4W  DowntownToronto   
11        M4X  DowntownToronto   
12        M4Y  DowntownToronto   
13        M5A  DowntownToronto   
14        M5B  DowntownToronto   
15        M5C  DowntownToronto   
16        M5E  DowntownToronto   
17        M5G  DowntownToronto   
18        M5H  DowntownToronto   
19        M5J  DowntownToronto   
20        M5K  DowntownToronto   
21        M5L  DowntownToronto   
22        M5N   CentralToronto   
23        M5P   CentralToronto   
24        M5R   CentralToronto   
25        M5S  DowntownToronto   
26        M5T  DowntownToronto   
27        M5V  DowntownToronto   
28        M5W  DowntownToronto   
29        M5X  DowntownToronto   
30        M6G  DowntownToronto   
31        M6H      WestToronto   
32        M6J      WestToronto   
33        M6K      WestToronto   
34        M6P      WestToronto   
35        M6R      WestToronto   
36        M6S      WestToronto   
37        M7A  DowntownToronto   
38        M7Y      EastToronto   

                                        Neighborhoods   Airport  \
0                                          TheBeaches  0.000000   
1                           TheDanforthWest,Riverdale  0.000000   
2                          IndiaBazaar,TheBeachesWest  0.000000   
3                                      StudioDistrict  0.000000   
4                                        LawrencePark  0.000000   
5                                     DavisvilleNorth  0.000000   
6                       NorthTorontoWest,LawrencePark  0.000000   
7                                          Davisville  0.000000   
8                            MoorePark,SummerhillEast  0.000000   
9   SummerhillWest,Rathnelly,SouthHill,ForestHillS...  0.000000   
10                                           Rosedale  0.000000   
11                           St.JamesTown,Cabbagetown  0.000000   
12                                 ChurchandWellesley  0.000000   
13                            RegentPark,Harbourfront  0.000000   
14                             GardenDistrict,Ryerson  0.000000   
15                                       St.JamesTown  0.000000   
16                                         BerczyPark  0.000000   
17                                   CentralBayStreet  0.000000   
18                             Richmond,Adelaide,King  0.000000   
19       HarbourfrontEast,UnionStation,TorontoIslands  0.000000   
20               TorontoDominionCentre,DesignExchange  0.000000   
21                        CommerceCourt,VictoriaHotel  0.000000   
22                                           Roselawn  0.000000   
23            ForestHillNorth&West,ForestHillRoadPark  0.000000   
24                    TheAnnex,NorthMidtown,Yorkville  0.000000   
25                        UniversityofToronto,Harbord  0.000000   
26              KensingtonMarket,Chinatown,GrangePark  0.000000   
27  CNTower,KingandSpadina,RailwayLands,Harbourfro...  0.071429   
28                                        StnAPOBoxes  0.000000   
29                 FirstCanadianPlace,Undergroundcity  0.000000   
30                                           Christie  0.000000   
31                         Dufferin,DovercourtVillage  0.000000   
32                             LittlePortugal,Trinity  0.000000   
33           Brockton,ParkdaleVillage,ExhibitionPlace  0.000000   
34                          HighPark,TheJunctionSouth  0.000000   
35                              Parkdale,Roncesvalles  0.000000   
36                                  Runnymede,Swansea  0.000000   
37            Queen'sPark,OntarioProvincialGovernment  0.000000   
38  BusinessreplymailProcessi

Now let's create the new dataframe and display the top 10 venues for each PostalCode.

In [55]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
areaColumns = ['Postal Code', 'Borough', 'Neighborhoods']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most Common Venue'.format(ind+1))
columns = areaColumns+freqColumns

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Postal Code'] = toronto_grouped['Postal Code']
neighborhoods_venues_sorted['Borough'] = toronto_grouped['Borough']
neighborhoods_venues_sorted['Neighborhoods'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    row_categories = toronto_grouped.iloc[ind, :].iloc[3:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    neighborhoods_venues_sorted.iloc[ind, 3:] = row_categories_sorted.index.values[0:num_top_venues]

# neighborhoods_venues_sorted.sort_values(freqColumns, inplace=True)
print(neighborhoods_venues_sorted.shape)
neighborhoods_venues_sorted

KeyError: 'Postal Code'

# Cluster Neighbourhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [56]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop(["PostalCode", "Borough", "Neighborhoods"], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([4, 1, 1, 1, 2, 1, 1, 1, 1, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [57]:
toronto_merged=df

# add clustering labels
toronto_merged["Cluster Labels"] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.drop(["Borough", "Neighborhoods"], 1).set_index("Postal Code"), on="Postal Code")

print(toronto_merged.shape)
toronto_merged.head() # check the last columns!

(39, 16)


Postal Code         Borough                Neighborhood   Latitude  \
0         M4E     EastToronto                  TheBeaches  43.676357   
1         M4K     EastToronto   TheDanforthWest,Riverdale  43.679557   
2         M4L     EastToronto  IndiaBazaar,TheBeachesWest  43.668999   
3         M4M     EastToronto              StudioDistrict  43.659526   
4         M4N  CentralToronto                LawrencePark  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               4                   NaN                   NaN   
1 -79.352188               1                   NaN                   NaN   
2 -79.315572               1                   NaN                   NaN   
3 -79.340923               1                   NaN                   NaN   
4 -79.388790               2                   NaN                   NaN   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   NaN                   NaN                   NaN   
1                   NaN                   NaN                   NaN   
2                   NaN                   NaN                   NaN   
3                   NaN                   NaN                   NaN   
4                   NaN                   NaN                   NaN   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                   NaN                   NaN                   NaN   
1                   NaN                   NaN                   NaN   
2                   NaN                   NaN                   NaN   
3                   NaN                   NaN                   NaN   
4                   NaN                   NaN                   NaN   

  9th Most Common Venue 10th Most Common Venue  
0                   NaN                    NaN  
1                   NaN                    NaN  
2                   NaN                    NaN  
3                   NaN                    NaN  
4                   NaN                    NaN

Finally, let's visualize the resulting clusters

In [58]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examine the clusters separately

### Cluster 1

In [63]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
23  CentralToronto               0    Mexican Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
23                 Trail         Jewelry Store      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
23           Yoga Studio                 Diner    Falafel Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
23           Event Space  Ethiopian Restaurant      Electronics Store

### Cluster 2

In [64]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1       EastToronto               1      Greek Restaurant   
2       EastToronto               1          Intersection   
3       EastToronto               1                  Café   
5    CentralToronto               1                 Hotel   
6    CentralToronto               1        Clothing Store   
7    CentralToronto               1          Dessert Shop   
8    CentralToronto               1                 Trail   
9    CentralToronto               1           Coffee Shop   
11  DowntownToronto               1           Coffee Shop   
12  DowntownToronto               1           Coffee Shop   
13  DowntownToronto               1           Coffee Shop   
14  DowntownToronto               1        Clothing Store   
15  DowntownToronto               1           Coffee Shop   
16  DowntownToronto               1           Coffee Shop   
17  DowntownToronto               1           Coffee Shop   
18  DowntownToronto               1           Coffee Shop   
19  DowntownToronto               1           Coffee Shop   
20  DowntownToronto               1           Coffee Shop   
21  DowntownToronto               1           Coffee Shop   
24   CentralToronto               1        Sandwich Place   
25  DowntownToronto               1                  Café   
26  DowntownToronto               1                  Café   
27  DowntownToronto               1       Airport Service   
28  DowntownToronto               1           Coffee Shop   
29  DowntownToronto               1           Coffee Shop   
30  DowntownToronto               1         Grocery Store   
31      WestToronto               1              Pharmacy   
32      WestToronto               1                   Bar   
33      WestToronto               1                  Café   
34      WestToronto               1    Mexican Restaurant   
35      WestToronto               1        Breakfast Spot   
36      WestToronto               1                  Café   
37  DowntownToronto               1           Coffee Shop   
38      EastToronto               1    Light Rail Station   

   2nd Most Common Venue  3rd Most Common Venue  \
1            Coffee Shop     Italian Restaurant   
2         Ice Cream Shop                Brewery   
3            Coffee Shop                 Bakery   
5                   Park      Food & Drink Shop   
6    Sporting Goods Shop            Coffee Shop   
7         Sandwich Place       Sushi Restaurant   
8                   Park             Restaurant   
9                    Pub           Liquor Store   
11                  Café                 Bakery   
12      Sushi Restaurant    Japanese Restaurant   
13                Bakery                   Park   
14           Coffee Shop    Japanese Restaurant   
15                  Café             Restaurant   
16          Cocktail Bar                 Bakery   
17    Italian Restaurant         Sandwich Place   
18                  Café                  Hotel   
19              Aquarium                   Café   
20                 Hotel                   Café   
21                  Café             Restaurant   
24                  Café            Coffee Shop   
25                   Bar                 Bakery   
26           Coffee Shop  Vietnamese Restaurant   
27               Airport          Boat or Ferry   
28                  Café                  Hotel   
29                  Café                  Hotel   
30                  Café                   Park   
31                Bakery                   Park   
32                  Café  Vietnamese Restaurant   
33           Coffee Shop         Breakfast Spot   
34                  Café        Thai Restaurant   
35             Gift Shop       Cuban Restaurant   
36           Coffee Shop              Bookstore   
37                 Diner            Yoga Studio   
38        Farmers Market                 Garden   

            4th Most Common Venue   5th Most Common Venue  \
1                  Ice Cream 

### Cluster 3

In [66]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
4    CentralToronto               2                  Park   
10  DowntownToronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4               Bus Line           Swim School           Yoga Studio   
10            Playground                 Trail           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4         Discount Store        Farmers Market    Falafel Restaurant   
10          Dessert Shop    Falafel Restaurant           Event Space   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
4            Event Space  Ethiopian Restaurant            Electronics Store  
10  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant

### Cluster 4

In [68]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  CentralToronto               3                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22          Home Service           Yoga Studio        Farmers Market   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22    Falafel Restaurant           Event Space  Ethiopian Restaurant   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
22     Electronics Store  Eastern European Restaurant             Donut Shop

### Cluster 5

In [69]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  EastToronto               4          Neighborhood     Health Food Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pub                 Trail           Event Space   

  6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0      Department Store             Donut Shop

# Observations :

Most of the venues fall in Cluster 2 Followed by cluster 3,1,4 and 5.

Now we will try to find clusters according to the number of shopping malls in the neighburhoods so we can easily determine the prime location for setting up the shopping mall in Toronto.

In [59]:
len(toronto_grouped[toronto_grouped["Shopping Mall"] > 0])

5

### Create a new DataFrame for Shopping Mall data only

In [60]:
toronto_mall = toronto_grouped[["Neighborhoods","Shopping Mall"]]

In [61]:
toronto_mall.head()

Neighborhoods  Shopping Mall
0                  TheBeaches            0.0
1   TheDanforthWest,Riverdale            0.0
2  IndiaBazaar,TheBeachesWest            0.0
3              StudioDistrict            0.0
4                LawrencePark            0.0

Now we cluster toronto into 3 clusters showing the presence of shopping malls in various clusters.

In [64]:
# set number of clusters
kclusters = 3

toronto_clustering = toronto_mall.drop(["Neighborhoods"], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Add the clustering label to the dataframe of the malls in Toronto to see in which cluster the neighborhood belongs to.

In [66]:
# create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.
toronto_merged =toronto_mall.copy()

# add clustering labels
toronto_merged["Cluster Labels"] = kmeans.labels_

In [77]:
toronto_merged.rename(columns={'Neighborhoods':'Neighborhood'},inplace=True)
toronto_merged.head()

Neighborhood  Shopping Mall  Cluster Labels Postal Code  \
0                  TheBeaches            0.0               0         M4E   
1   TheDanforthWest,Riverdale            0.0               0         M4K   
2  IndiaBazaar,TheBeachesWest            0.0               0         M4L   
3              StudioDistrict            0.0               0         M4M   
4                LawrencePark            0.0               0         M4N   

          Borough  
0     EastToronto  
1     EastToronto  
2     EastToronto  
3     EastToronto  
4  CentralToronto

In [80]:
# sort the results by Cluster Labels
print(toronto_merged.shape)
toronto_merged.sort_values(["Cluster Labels"], inplace=True)
toronto_merged

(39, 5)


Neighborhood  Shopping Mall  \
0                                          TheBeaches       0.000000   
22                                           Roselawn       0.000000   
23            ForestHillNorth&West,ForestHillRoadPark       0.000000   
24                    TheAnnex,NorthMidtown,Yorkville       0.000000   
25                        UniversityofToronto,Harbord       0.000000   
26              KensingtonMarket,Chinatown,GrangePark       0.000000   
27  CNTower,KingandSpadina,RailwayLands,Harbourfro...       0.000000   
37            Queen'sPark,OntarioProvincialGovernment       0.000000   
29                 FirstCanadianPlace,Undergroundcity       0.000000   
31                         Dufferin,DovercourtVillage       0.000000   
32                             LittlePortugal,Trinity       0.000000   
33           Brockton,ParkdaleVillage,ExhibitionPlace       0.000000   
34                          HighPark,TheJunctionSouth       0.000000   
35                              Parkdale,Roncesvalles       0.000000   
36                                  Runnymede,Swansea       0.000000   
30                                           Christie       0.000000   
18                             Richmond,Adelaide,King       0.000000   
19       HarbourfrontEast,UnionStation,TorontoIslands       0.000000   
38  BusinessreplymailProcessingCentre,SouthCentral...       0.000000   
15                                       St.JamesTown       0.000000   
1                           TheDanforthWest,Riverdale       0.000000   
2                          IndiaBazaar,TheBeachesWest       0.000000   
3                                      StudioDistrict       0.000000   
13                            RegentPark,Harbourfront       0.000000   
12                                 ChurchandWellesley       0.000000   
11                           St.JamesTown,Cabbagetown       0.000000   
10                                           Rosedale       0.000000   
9   SummerhillWest,Rathnelly,SouthHill,ForestHillS...       0.000000   
17                                   CentralBayStreet       0.000000   
8                            MoorePark,SummerhillEast       0.000000   
7                                          Davisville       0.000000   
6                       NorthTorontoWest,LawrencePark       0.000000   
5                                     DavisvilleNorth       0.000000   
4                                        LawrencePark       0.000000   
28                                        StnAPOBoxes       0.010309   
21                        CommerceCourt,VictoriaHotel       0.010000   
20               TorontoDominionCentre,DesignExchange       0.010000   
14                             GardenDistrict,Ryerson       0.010000   
16                                         BerczyPark       0.017241   

    Cluster Labels Postal Code          Borough  
0                0         M4E      EastToronto  
22               0         M5N   CentralToronto  
23               0         M5P   CentralToronto  
24               0         M5R   CentralToronto  
25               0         M5S  DowntownToronto  
26               0         M5T  DowntownToronto  
27               0         M5V  DowntownToronto  
37               0         M7A  DowntownToronto  
29               0         M5X  DowntownToronto  
31               0         M6H      WestToronto  
32               0         M6J      WestToronto  
33               0         M6K      WestToronto  
34               0         M6P      WestToronto  
35               0         M6R      WestToronto  
36               0         M6S      WestToronto  
30               0         M6G  DowntownToronto  
18               0         M5H  DowntownToronto  
19               0         M5J  DowntownToronto  
38               0         M7Y      EastToronto  
15               0         M5C  DowntownToronto  
1                0         M4K      EastToronto  
2                0         M4L      EastToronto  
3         

## Now finally we examine the individual clusters for the shopping malls.

#### Cluster 1

In [81]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0]

Neighborhood  Shopping Mall  \
0                                          TheBeaches            0.0   
22                                           Roselawn            0.0   
23            ForestHillNorth&West,ForestHillRoadPark            0.0   
24                    TheAnnex,NorthMidtown,Yorkville            0.0   
25                        UniversityofToronto,Harbord            0.0   
26              KensingtonMarket,Chinatown,GrangePark            0.0   
27  CNTower,KingandSpadina,RailwayLands,Harbourfro...            0.0   
37            Queen'sPark,OntarioProvincialGovernment            0.0   
29                 FirstCanadianPlace,Undergroundcity            0.0   
31                         Dufferin,DovercourtVillage            0.0   
32                             LittlePortugal,Trinity            0.0   
33           Brockton,ParkdaleVillage,ExhibitionPlace            0.0   
34                          HighPark,TheJunctionSouth            0.0   
35                              Parkdale,Roncesvalles            0.0   
36                                  Runnymede,Swansea            0.0   
30                                           Christie            0.0   
18                             Richmond,Adelaide,King            0.0   
19       HarbourfrontEast,UnionStation,TorontoIslands            0.0   
38  BusinessreplymailProcessingCentre,SouthCentral...            0.0   
15                                       St.JamesTown            0.0   
1                           TheDanforthWest,Riverdale            0.0   
2                          IndiaBazaar,TheBeachesWest            0.0   
3                                      StudioDistrict            0.0   
13                            RegentPark,Harbourfront            0.0   
12                                 ChurchandWellesley            0.0   
11                           St.JamesTown,Cabbagetown            0.0   
10                                           Rosedale            0.0   
9   SummerhillWest,Rathnelly,SouthHill,ForestHillS...            0.0   
17                                   CentralBayStreet            0.0   
8                            MoorePark,SummerhillEast            0.0   
7                                          Davisville            0.0   
6                       NorthTorontoWest,LawrencePark            0.0   
5                                     DavisvilleNorth            0.0   
4                                        LawrencePark            0.0   

    Cluster Labels Postal Code          Borough  
0                0         M4E      EastToronto  
22               0         M5N   CentralToronto  
23               0         M5P   CentralToronto  
24               0         M5R   CentralToronto  
25               0         M5S  DowntownToronto  
26               0         M5T  DowntownToronto  
27               0         M5V  DowntownToronto  
37               0         M7A  DowntownToronto  
29               0         M5X  DowntownToronto  
31               0         M6H      WestToronto  
32               0         M6J      WestToronto  
33               0         M6K      WestToronto  
34               0         M6P      WestToronto  
35               0         M6R      WestToronto  
36               0         M6S      WestToronto  
30               0         M6G  DowntownToronto  
18               0         M5H  DowntownToronto  
19               0         M5J  DowntownToronto  
38               0         M7Y      EastToronto  
15               0         M5C  DowntownToronto  
1                0         M4K      EastToronto  
2                0         M4L      EastToronto  
3                0         M4M      EastToronto  
13               0         M5A  DowntownToronto  
12               0         M4Y  DowntownToronto  
11               0         M4X  DowntownToronto  
10               0         M4W  DowntownToronto  
9                0         M4V   CentralToronto  
17               0         M5G  DowntownToronto  
8                0  

#### Cluster 2

In [82]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1]

Neighborhood  Shopping Mall  Cluster Labels  \
28                           StnAPOBoxes       0.010309               1   
21           CommerceCourt,VictoriaHotel       0.010000               1   
20  TorontoDominionCentre,DesignExchange       0.010000               1   
14                GardenDistrict,Ryerson       0.010000               1   

   Postal Code          Borough  
28         M5W  DowntownToronto  
21         M5L  DowntownToronto  
20         M5K  DowntownToronto  
14         M5B  DowntownToronto

#### Cluster 3

In [83]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2]

Neighborhood  Shopping Mall  Cluster Labels Postal Code          Borough
16   BerczyPark       0.017241               2         M5E  DowntownToronto

## Observations :

Most of the shopping malls are present in Cluster 1 while cluster 2 has very low number and cluster 3 almost has no shopping malls.Thus, shopping malls in cluster 1 are likely suffering from intense competition due to oversupply and high concentration of shopping malls.Thus from the competition point of view developers should open a mall either in cluster 2 or in cluster 3 avoiding stiff competition.